design/design-generic-references.md

```diff
@@ -0,0 +1,314 @@
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      name: main-vpc
+      fieldPath: status.atProvider.id
+status:
+  value: foo-vpc-id # the value found in the requested field
+  conditions:
+    - type: Ready
+      status: True
+      reason: Available
+```
+
+Now, the value can be referred in the usual way within a managed resource:
+
+```yaml
+spec:
+  vpcIdRef: referable-vpc-id
+```
+
+The reference resolver could first try to find `referable-vpc-id` `VPC` instance, 
+and fallback to `referable-vpc-id` `Referable` if the former does not exist. 
+If the instance is ready, the value is read from `status.value` and 
+set to `spec.vpcId` of the managed resource. Ofcourse, if needed, and if it 
+makes more sense, the resolver could first look for the existence of `Referable` 
+object first.
+
+Referring a namespaced object would be possible as well:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: region
+spec:
+  objects:
+    - apiVersion: v1
+      kind: ConfigMap
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+```
+
+or we could match an object with a label selector, if its name is unknown or not
+static:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: referable-vpc-id
+spec:
+  objects:
+    - apiVersion: ec2.aws.crossplane.io/v1alpha1
+      kind: VPC
+      matchingLabels:
+        class: main-vpc
+      fieldPath: status.atProvider.id
+```
+
+### Value Assigning Without Counterpart Ref Field
+
+If a value needs to be assigned to a field that does not have
+corresponding `Ref` field, the reference can be expressed using `spec.refs` block:
+
+```yaml
+spec:
+  refs:
+    - name: referable-vpc-id
+      toFieldPath: spec.myVPCId
+```
+
+The syntax of `toFieldPath` field is similar to what we
+use in Composition, in line [Kubernetes API Conventions on field selection].
+
+Alternatively, if we would like to avoid changing managed resource schema, the
+above can be stated using annotations as well:
+
+```yaml
+metadata:
+  annotations:
+    "referable.upbound.io/referable-vpc-id": "spec.myVPCId"
+```
+
+### Referring Value Existing Potentially in Multiple Source
+
+If a value could be found in multiple objects of different kinds, we can mark
+them as optional and pick the value from the first one found:
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: Referable
+metadata:
+  name: multi-ref
+spec:
+  objects:
+    - apiVersion: example.com/v1
+      kind: Foo
+      name: foo
+      optional: true
+      fieldPath: spec.id
+    - apiVersion: example.com/v1
+      kind: Bar
+      name: bar
+      optional: true
+      fieldPath: spec.myId
+```
+
+The above syntax can be used as well when a value might appear in a several
+places within single object.
+
+### Value Transformation
```

Comment:
It might be useful to refer to a real-world example that might lead to this scenario.

Reply:
sure, @turkenh or @haarchri do you have some better example from the field?

Reply:
Yes, a more complex example would be nice to see, not necessarily a real world example, but at least a bigger toy example would be helpful. I'm not sure how you are envisioning this new API to be used in the context of Compositions. Should users define this new resource as part of their Composition? How will that look like? So users will most probably have to define `Usage` resources as defined by https://github.com/crossplane/crossplane/pull/4215 to achieve deletion ordering and then this new object to define mappings between values? how's that going to look like?